includes/amp-helper-functions.php

```diff
@@ -299,6 +299,85 @@ function amp_is_canonical() {
  */
 function amp_add_frontend_actions() {
 	add_action( 'wp_head', 'amp_add_amphtml_link' );
+
+	if ( AMP_Options_Manager::get_option( Option::MOBILE_REDIRECT ) && amp_version_is_available() ) {
+		// Insert the mobile redirect script as early as possible.
+		add_action( 'wp_head', 'amp_add_mobile_redirect_script', ~PHP_INT_MAX );
+	}
+}
+
+/**
+ * Add mobile redirect script.
+ *
+ * @since 1.6
+ */
+function amp_add_mobile_redirect_script() {
+	?>
+	<script>
+		(function ( ampSlug ) {
+			var hasTouchScreen;
+
+			if ("maxTouchPoints" in navigator) {
+				hasTouchScreen = navigator.maxTouchPoints > 0;
+			} else if ("msMaxTouchPoints" in navigator) {
+				hasTouchScreen = navigator.msMaxTouchPoints > 0;
+			} else {
+				var mQ = window.matchMedia && matchMedia("(pointer:coarse)");
+				if (mQ && mQ.media === "(pointer:coarse)") {
+					hasTouchScreen = !!mQ.matches;
+				} else if ('orientation' in window) {
+					hasTouchScreen = true; // deprecated, but good fallback
+				} else {
+					// Only as a last resort, fall back to user agent sniffing
+					var UA = navigator.userAgent;
+					hasTouchScreen = /\b(Mobile|Android|Silk\/|Kindle|BlackBerry|Opera Mini|Opera Mobi|webOS|iPhone|IEMobile|Windows Phone|iPad|iPod)\b/.test( UA );
+				}
+			}
+
+			var url = new URL( location.href );
+
+			if ( hasTouchScreen && ! url.searchParams.has( ampSlug ) ) {
+				window.stop(); // Stop loading the page! This should cancel all loading resources.
+
+				// Replace the current page with the AMP version.
+				url.searchParams.append( ampSlug, '1' );
+				location.replace( url.href );
+			}
+		})(
+			<?php echo wp_json_encode( amp_get_slug() ); ?>
+		)
+	</script>
+	<?php
+}
+
+/**
+ * Determine whether there is an AMP version available for this URL.
+ *
+ * @since 1.6
+ * @global WP_Query $wp_query
+ *
+ * @return bool True if there is an AMP version, false otherwise.
+ */
+function amp_version_is_available() {
```

Comment:
See #4619 and #4620

Reply:
Would you consider having #4620 complete first, or port over the changes from that PR to here?

Reply:
I'm happy if you want to incorporate the changes here. I've closed that PR.